File: scripts/method_script.py
import esprima
import ast  # To safely parse the string representation of dictionaries
import csv
import sys


def parse_js_code(file_path):
    """Parses JavaScript code from a file."""
    try:
        with open(file_path, 'r') as file:
            code = file.read()
        # Remove $ExpectError comments
        code = code.replace('$ExpectError', '')
        print("Debug: JavaScript code being parsed:\n", code)  # Debug statement
        parsed = esprima.parse(code, loc=True, sourceType="module")
        return code, parsed
    except Exception as e:
        raise ValueError(f"Failed to parse JavaScript file: {e}")


def extract_test_block(node, target_line, code):
    """Extracts the test block containing the target line."""
    if hasattr(node, 'loc'):
        start_line = node.loc.start.line
        end_line = node.loc.end.line
        if start_line <= target_line <= end_line:
            if node.type == 'CallExpression':
                if hasattr(node.callee, 'name') and node.callee.name in [
                    'it', 'test', 'beforeEach', 'afterEach', 'xit', 'xdescribe', 'fdescribe'
                ]:
                    return format_test_block(code, start_line, end_line)
                elif node.callee.type == 'MemberExpression':
                    object_name = getattr(node.callee.object, 'name', None)
                    property_name = getattr(node.callee.property, 'name', None)
                    if object_name == 'describe' and property_name in ['only', 'skip']:
                        return format_test_block(code, start_line, end_line)

            if node.type == 'VariableDeclarator' and hasattr(node, 'init') and node.init:
                if node.init.type in ['ArrowFunctionExpression', 'FunctionExpression']:
                    return format_test_block(code, start_line, end_line)
                elif node.init.type == 'CallExpression':
                    callee = node.init.callee
                    object_name = getattr(callee.object, 'name', None)
                    property_name = getattr(callee.property, 'name', None)
                    if object_name == 'jest' and property_name == 'spyOn':
                        return format_test_block(code, start_line, end_line)

    for child in traverse_ast(node):
        result = extract_test_block(child, target_line, code)
        if result:
            return result
    return None


def format_test_block(code, start_line, end_line):
    """Formats the test block for output."""
    test_code = code.splitlines()[start_line - 1:end_line]
    return {
        "test": "\n".join(test_code),
        "start": start_line,
        "end": end_line
    }


def traverse_ast(node):
    """Traverses the AST recursively."""
    for key in node.__dict__:
        child = getattr(node, key)
        if isinstance(child, list):
            for sub_child in child:
                if isinstance(sub_child, esprima.nodes.Node):
                    yield sub_child
        elif isinstance(child, esprima.nodes.Node):
            yield child


def find_parent_test_method(file_path, target_line):
    """Finds the parent test method containing the target line."""
    try:
        code, parsed = parse_js_code(file_path)
        for node in parsed.body:
            result = extract_test_block(node, target_line, code)
            if result:
                return result
        return {"error": "No test method found containing the specified line."}
    except FileNotFoundError:
        return {"error": f"File not found: {file_path}"}
    except ValueError as e:
        return {"error": f"Parsing error: {e}"}
    except Exception as e:
        return {"error": f"Unexpected error: {e}"}


def process_csv(input_csv, output_csv):
    """Processes the input CSV and writes results to the output CSV."""
    with open(input_csv, 'r') as infile, open(output_csv, 'w', newline='') as outfile:
        reader = csv.DictReader(infile)
        fieldnames = ['A (filename)', 'B (test)', 'C (start: ... end: ....)']
        writer = csv.DictWriter(outfile, fieldnames=fieldnames)
        writer.writeheader()

        for row in reader:
            file_path, target_line = construct_file_path_and_line(row)
            if not file_path or not target_line:
                writer.writerow({
                    'A (filename)': file_path or "Invalid file path",
                    'B (test)': "Error: Invalid 'Lines' format",
                    'C (start: ... end: ....)': ''
                })
                continue

            result = find_parent_test_method(file_path, target_line)
            if "error" in result:
                writer.writerow({
                    'A (filename)': file_path,
                    'B (test)': result['error'],
                    'C (start: ... end: ....)': ''
                })
            else:
                writer.writerow({
                    'A (filename)': file_path,
                    'B (test)': result['test'],  # Include the original method code here
                    'C (start: ... end: ....)': f"start: {result['start']} end: {result['end']}"
                })


def construct_file_path_and_line(row):
    """Constructs the file path and extracts the target line from the CSV row."""
    try:
        print("Debug: Processing row:", row)  # Debug statement
        repo_name = row['Repository'].split('/')[-1]
        project_path = "/home/gabriel/Desktop/research/projects/"
        file_path = f"{project_path}{repo_name}{row['File']}"

        lines_data = ast.literal_eval(row['Lines'])
        target_line = lines_data.get('startLine') or lines_data.get('line')
        return file_path, target_line
    except (ValueError, SyntaxError, KeyError) as e:
        print("Debug: Error processing row:", e)  # Debug statement
        return None, None


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python method_script.py <input_csv> <output_csv>")
        sys.exit(1)

    input_csv = sys.argv[1]
    output_csv = sys.argv[2]
    process_csv(input_csv, output_csv)

File: scripts/method_script/method_script.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');
const { parse } = require('@babel/parser');
const traverse = require('@babel/traverse').default;
const { stringify } = require('csv-stringify/sync');

/**
 * Constructs the file path and extracts the target line from the CSV row.
 * @param {object} row - A row from the CSV file.
 * @returns {object} - An object containing the file path and target line.
 */
function constructFilePathAndLine(row) {
  try {
    console.log('Debug: Processing row:', row);

    // Validate Repository and File fields
    if (!row['Repository'] || !row['File']) {
      throw new Error('Missing "Repository" or "File" field in the row.');
    }

    const repoName = row['Repository'].split('/').pop();

    // Use the PROJECT_PATH environment variable or fall back to a default value
    const projectPath = process.env.PROJECT_PATH || '/home/username/Desktop/research/projects/';
    if (!fs.existsSync(projectPath)) {
      throw new Error(`Invalid PROJECT_PATH: ${projectPath}`);
    }

    const filePath = path.join(projectPath, repoName, row['File']);

    // Validate and parse the Lines field
    if (!row['Lines']) {
      throw new Error('Missing "Lines" field in the row.');
    }

    const linesData = JSON.parse(row['Lines'].replace(/'/g, '"'));
    const targetLine = linesData.startLine || linesData.line;

    if (!targetLine || typeof targetLine !== 'number') {
      throw new Error('Invalid "Lines" field format.');
    }

    return { filePath, targetLine };
  } catch (error) {
    console.error('Debug: Error processing row:', error.message);
    return { filePath: null, targetLine: null };
  }
}

/**
 * Extracts the test method containing the target line from a JavaScript file.
 * @param {string} filePath - Path to the JavaScript file.
 * @param {number} targetLine - Line number to locate the test method.
 * @returns {object} - Extracted test method or an error message.
 */
function findParentTestMethod(filePath, targetLine) {
  try {
    if (!fs.existsSync(filePath)) {
      throw new Error(`File not found: ${filePath}`);
    }

    const code = fs.readFileSync(filePath, 'utf-8');
    const ast = parse(code, {
      sourceType: 'module',
      plugins: ['flow', 'jsx'], // Enable Flow and JSX parsing
    });

    let result = { error: 'No test method found containing the specified line.' };

    traverse(ast, {
      CallExpression(path) {
        const { node } = path;

        // Check for test methods like `it`, `test`, `xit`, `fit`
        if (
          node.callee.type === 'Identifier' &&
          ['it', 'test', 'beforeEach', 'afterEach', 'xit', 'xdescribe', 'fdescribe'].includes(node.callee.name)
        ) {
          const startLine = node.loc.start.line;
          const endLine = node.loc.end.line;

          if (startLine <= targetLine && targetLine <= endLine) {
            result = {
              test: code.split('\n').slice(startLine - 1, endLine).join('\n'),
              start: startLine,
              end: endLine,
            };
          }
        }

        // Check for `describe.only` or `describe.skip`
        if (
          node.callee.type === 'MemberExpression' &&
          node.callee.object.type === 'Identifier' &&
          node.callee.object.name === 'describe' &&
          ['only', 'skip'].includes(node.callee.property.name)
        ) {
          const startLine = node.loc.start.line;
          const endLine = node.loc.end.line;

          if (startLine <= targetLine && targetLine <= endLine) {
            result = {
              test: code.split('\n').slice(startLine - 1, endLine).join('\n'),
              start: startLine,
              end: endLine,
            };
          }
        }
      },
    });

    return result;
  } catch (error) {
    return { error: `Failed to parse JavaScript file: ${error.message}` };
  }
}

/**
 * Processes the input CSV and writes results to the output CSV.
 * @param {string} inputCsv - Path to the input CSV file.
 * @param {string} outputCsv - Path to the output CSV file.
 */
function processCsv(inputCsv, outputCsv) {
  const rows = [];
  fs.createReadStream(inputCsv)
    .pipe(csv())
    .on('data', (row) => rows.push(row))
    .on('end', () => {
      const results = rows.map((row) => {
        const { filePath, targetLine } = constructFilePathAndLine(row);

        if (!filePath || !targetLine) {
          return {
            'filename': filePath || 'Invalid file path',
            'test': 'Error: Invalid "Lines" format',
            'start: ... end: ...': '',
          };
        }

        const result = findParentTestMethod(filePath, targetLine);
        if (result.error) {
          return {
            'filename': filePath,
            'test': result.error,
            'start: ... end: ...': '',
          };
        }

        return {
          'filename': filePath,
          'test': result.test,
          'start: ... end: ...': `start: ${result.start} end: ${result.end}`,
        };
      });

      const csvOutput = stringify(results, {
        header: true,
        columns: ['filename', 'test', 'start: ... end: ...'],
      });

      fs.writeFileSync(outputCsv, csvOutput);
      console.log(`Results written to ${outputCsv}`);
    })
    .on('error', (error) => {
      console.error(`Error reading input CSV: ${error.message}`);
    });
}

// Main script execution
if (require.main === module) {
  const inputCsv = process.argv[2];
  const outputCsv = process.argv[3];

  if (!inputCsv || !outputCsv) {
    console.error('Usage: node method_script.js <input_csv> <output_csv>');
    process.exit(1);
  }

  processCsv(inputCsv, outputCsv);
}